### PR TITLE
Add tagging and other support for singularity pull

### PR DESCRIPTION
**Pull request by dtrudg**
**Description of the Pull Request (PR):**

This PR adds CLI support for tagging in the Container library.

You can push an image with:
- no tag (will default to tagging with 'latest'
    `singularity push myimage.sif library://dave/test/alpine`
- a single tag
   `singularity push myimage.sif library://dave/test/alpine:demo1`
- multiple tags
  `singularity push myimage.sif library://dave/test/alpine:demo1,latest,meoww`

Images can be pulled with a named tag, or their id (currently in `sha256.<sha256>` format, but native `sif.<uuid>` to be added later)

Also:
- now uses sylog for log messages
- pull won't overwrite existing image file unless `--force/-F` specified
- can pull without specifying an image name - it'll use a default `<container>_<tag>.sif` format

### CODE REVIEW

File: src/cmd/singularity/cli/pull.go
Comment (GodloveD) [inline src/cmd/singularity/cli/pull.go]: Just a comment.  In general I think that we should push the runscripts elsewhere so that the CLI remains the CLI and they just call out to another function.  But there are several other instances of this that also need to be worked on.  I won't block b/c of it.  Just a comment.

File: src/pkg/library/client/push.go
Comment (GodloveD) [inline src/pkg/library/client/push.go:165]: Don't you need some documentation above functions or some such thing?

**Review by GodloveD — DISMISSED** [review]
If you want a thorough review on this it will probably take me several days.  If you want this in quickly it may be best to get another reviewer I think.  I will approve this because I trust you and I know the code is necessarily rough and ready right now.  Up to you if you want to get a second review or if you need this in to keep making progress.

File: src/pkg/library/client/push.go
Comment (dtrudg) [reply src/pkg/library/client/push.go:165]: The general rule (applied by go lint at least) is that public functions (starting with capitals) need doc comments, and private functions like this don't. But that wouldn't preclude me adding some more here.

**Comment by dtrudg** [comment]
@GodloveD - many thanks for this.  Hopefully @bauerm97 or @tri-adam can take a quick look through too, as it's a fair chunk.

Noting for this PR, also had useful discussion with Dave r.e. the library:// URI structure, and thoughts on naming for entities/collections etc. I will open an issue to revisit this and ensure that things are consistent in any user-exposed messages / interface.

**Review by tri-adam — DISMISSED** [review]
Looks good to me!

**Comment by dtrudg** [comment]
Hi @tri-adam @GodloveD  - merging development-3.0 back in dismissed your reviews. Are you still happy to merge this? The only change from the merge was on gopkg.lock - thanks.

**Review by tri-adam — APPROVED** [review]

**Pull request merged**
